tests/is-valid-utf8/Main.hs

```diff
@@ -0,0 +1,290 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Main (main) where
+
+import Data.Bits (shiftR, (.&.), shiftL)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import Data.Char (chr, ord)
+import Data.Word (Word8)
+import Test.QuickCheck (Property, forAll, (===))
+import Test.QuickCheck.Arbitrary (Arbitrary (arbitrary, shrink))
+import Test.QuickCheck.Gen (oneof, Gen, choose, vectorOf, listOf1, sized, resize,
+                            elements)
+import Test.Tasty (defaultMain, testGroup)
+import Test.Tasty.QuickCheck (testProperty)
+
+main :: IO ()
+main = defaultMain . testGroup "UTF-8 validation" $ [
+  testProperty "Valid UTF-8" goValid,
+  testProperty "Invalid UTF-8" goInvalid
```

Comment:
@kozross do `goValid` and `goInvalid` succeed in GHCi?

Reply:
`goValid` segfaults. `goInvalid` runs to completion successfully.